// @icedesign/img/lib/modules/Cover.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _aliCDNSuffix = _interopRequireDefault(require("./aliCDNSuffix"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var Cover =
/*#__PURE__*/
function (_Component) {
  _inherits(Cover, _Component);

  function Cover(props) {
    var _this;

    _classCallCheck(this, Cover);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Cover).call(this, props));

    _this.imgOnload = function () {
      _this.setState({
        imgStyles: {
          display: 'block',
          maxWidth: 'none',
          wdith: 'auto',
          height: 'auto',
          // transition: 'opacity 0.2s ease',
          opacity: 0
        }
      }, function () {
        var imgDOM = _this.refs.img;
        var _this$props = _this.props,
            width = _this$props.width,
            height = _this$props.height;
        var imgWidth = imgDOM.width;
        var imgHeight = imgDOM.height;
        var imgStyles = getStyles(width, height, imgWidth, imgHeight);

        _this.setState({
          imgStyles: _objectSpread({}, _this.state.imgStyles, imgStyles, {
            opacity: 1
          })
        });
      });
    };

    _this.state = {
      imgStyles: {
        display: 'block',
        maxWidth: 'none',
        wdith: 'auto',
        height: 'auto',
        // transition: 'opacity 0.2s ease',
        opacity: 0
      }
    };
    return _this;
  }

  _createClass(Cover, [{
    key: "componentWillReceiveProps",
    value: function componentWillReceiveProps(nextProps) {
      if (nextProps.width !== this.props.width || nextProps.height !== this.props.height) {
        this.imgOnload();
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          alt = _this$props2.alt,
          title = _this$props2.title,
          src = _this$props2.src,
          width = _this$props2.width,
          height = _this$props2.height,
          _this$props2$style = _this$props2.style,
          style = _this$props2$style === void 0 ? {} : _this$props2$style,
          className = _this$props2.className,
          onError = _this$props2.onError,
          shape = _this$props2.shape,
          enableAliCDNSuffix = _this$props2.enableAliCDNSuffix;

      var styles = _objectSpread({}, style, {
        display: 'inline-block',
        overflow: 'hidden',
        width: width,
        height: height
      });

      var imgStyles = this.state.imgStyles;
      var realSrc = enableAliCDNSuffix ? src + (0, _aliCDNSuffix.default)({
        width: width,
        height: height
      }) : src;
      var cls = (0, _classnames.default)('ice-img', shape, className);
      return _react.default.createElement("div", {
        className: cls,
        style: styles
      }, _react.default.createElement("img", {
        style: imgStyles,
        onError: onError,
        ref: "img",
        src: realSrc,
        alt: alt,
        title: title,
        onLoad: this.imgOnload
      }));
    }
  }]);

  return Cover;
}(_react.Component);

exports.default = Cover;
Cover.displayName = 'Cover';
Cover.propTypes = {
  src: _propTypes.default.string.isRequired,
  width: _propTypes.default.number,
  height: _propTypes.default.number,
  alt: _propTypes.default.string,
  style: _propTypes.default.object,
  onError: _propTypes.default.func
};
Cover.defaultProps = {
  onError: function onError() {},
  alt: ''
};

var getStyles = function getStyles(wrapWidth, wrapHeight, imgWidth, imgHeight) {
  // 进行等比运算
  var wrapRatio = wrapWidth / wrapHeight;
  var imgRatio = imgWidth / imgHeight;
  var finalWidth = 0;
  var finalHeight = 0;
  var finalMarginTop = 0;
  var finalMarginLeft = 0;

  if (wrapRatio > imgRatio) {
    // 父框宽度更大，以父框的宽度为准，拉伸图片上移
    finalWidth = wrapWidth;
    finalHeight = finalWidth / imgRatio;
    finalMarginTop = (finalHeight - wrapHeight) / 2 * -1;
  } else {
    // 父框高度更大，以父框高度为准，拉伸图片左移动
    finalHeight = wrapHeight;
    finalWidth = finalHeight * imgRatio;
    finalMarginLeft = (finalWidth - wrapWidth) / 2 * -1;
  }

  return {
    height: finalHeight,
    width: finalWidth,
    marginLeft: finalMarginLeft,
    marginTop: finalMarginTop
  };
};
